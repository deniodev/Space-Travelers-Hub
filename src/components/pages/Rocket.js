import React from 'react';

const Rocket = () => (
  <div>Rockets</div>
);

export default Rocket;
