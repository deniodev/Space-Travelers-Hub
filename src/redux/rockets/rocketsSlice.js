import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const initialState = {
  rockets: [],
  status: 'idle',
  isFetched: false,
  error: null,
};

export const fetchRockets = createAsyncThunk(
  'rockets/fetchRockets',
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/rockets');
    const data = await response.json();
    return data;
  },
);

export const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket(state, action) {
      const newRockets = state.rockets.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: true };
      });
      return {
        ...state,
        rockets: newRockets,
      };
    },
    cancelRocket(state, action) {
      const newRockets = state.rockets.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: false };
      });
      return {
        ...state,
        rockets: newRockets,
      };
    },
  },

  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => ({ ...state, status: 'loading' }))
      .addCase(fetchRockets.fulfilled, (state, action) => ({
        ...state, status: 'succeeded', rockets: action.payload, isFetched: true,
      }))
      .addCase(fetchRockets.rejected, (state, action) => ({ ...state, status: 'failed', error: action.error.message }));
  },
});

export default rocketsSlice.reducer;
export const { reserveRocket, cancelRocket } = rocketsSlice.actions;
