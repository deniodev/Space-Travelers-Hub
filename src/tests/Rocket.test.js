import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockstore from 'redux-mock-store';
import Rocket from '../components/pages/Rocket';
import '@testing-library/jest-dom';
import { reserveRocket } from '../redux/rockets/rocketsSlice';

const mockStore = configureMockstore([]);

describe('Rockets component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      rockets: {
        rockets: [
          {
            id: 1,
            name: 'Falcon 9',
            description: 'A reusable rocket developed by SpaceX',
            images: '#',
            reserved: false,
          },
          {
            id: 2,
            name: 'Atlas V',
            description: 'A rocket developed by United Launch Alliance',
            images: '#',
            reserved: false,
          },
        ],
        isFetched: true,
      },
    });
    store.dispatch(reserveRocket(1));
  });
  it('should render component with a list of rockets', () => {
    render(
      <Provider store={store}>
        <Rocket />
      </Provider>,
    );

    const headings = screen.queryAllByRole('heading', { name: /rockets/i });
    expect(headings.length).toBe(1);

    const rocketNames = screen.getAllByRole('heading', { level: 2 });
    expect(rocketNames.length).toBe(2);

    const rocketImages = screen.getAllByRole('img');
    expect(rocketImages.length).toBe(2);
  });

  it('test for Rockets component with button click', () => {
    const reservebtn = render(
      <Provider store={store}>
        <Rocket />
      </Provider>,
    );
    expect(reservebtn).toMatchInlineSnapshot(`
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Rockets
        </h1>
        <div
          class="rocket-list-item"
        >
          <img
            alt="Falcon 9"
            class="rocket-image"
          />
          <div
            class="rocket-info"
          >
            <h2 />
            <p>
               
              A reusable rocket developed by SpaceX
            </p>
            <button
              class="rocketreserved-btn"
              type="button"
            >
              Reserve rocket
            </button>
          </div>
        </div>
        <div
          class="rocket-list-item"
        >
          <img
            alt="Atlas V"
            class="rocket-image"
          />
          <div
            class="rocket-info"
          >
            <h2 />
            <p>
               
              A rocket developed by United Launch Alliance
            </p>
            <button
              class="rocketreserved-btn"
              type="button"
            >
              Reserve rocket
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Rockets
      </h1>
      <div
        class="rocket-list-item"
      >
        <img
          alt="Falcon 9"
          class="rocket-image"
        />
        <div
          class="rocket-info"
        >
          <h2 />
          <p>
             
            A reusable rocket developed by SpaceX
          </p>
          <button
            class="rocketreserved-btn"
            type="button"
          >
            Reserve rocket
          </button>
        </div>
      </div>
      <div
        class="rocket-list-item"
      >
        <img
          alt="Atlas V"
          class="rocket-image"
        />
        <div
          class="rocket-info"
        >
          <h2 />
          <p>
             
            A rocket developed by United Launch Alliance
          </p>
          <button
            class="rocketreserved-btn"
            type="button"
          >
            Reserve rocket
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`);
  });
});
