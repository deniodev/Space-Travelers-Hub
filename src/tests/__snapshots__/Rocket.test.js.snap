// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component test for Rockets component with button click 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Rockets
        </h1>
        <div
          class="rocket-list-item"
        >
          <img
            class="rocket-image"
          />
          <div
            class="rocket-info"
          >
            <h2>
              Falcon 9
            </h2>
            <p>
               
              A reusable rocket developed by SpaceX
            </p>
            <button
              class="rocketreserved-btn"
              type="button"
            >
              Reserve rocket
            </button>
          </div>
        </div>
        <div
          class="rocket-list-item"
        >
          <img
            class="rocket-image"
          />
          <div
            class="rocket-info"
          >
            <h2>
              Atlas V
            </h2>
            <p>
               
              A rocket developed by United Launch Alliance
            </p>
            <button
              class="rocketreserved-btn"
              type="button"
            >
              Reserve rocket
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Rockets
      </h1>
      <div
        class="rocket-list-item"
      >
        <img
          class="rocket-image"
        />
        <div
          class="rocket-info"
        >
          <h2>
            Falcon 9
          </h2>
          <p>
             
            A reusable rocket developed by SpaceX
          </p>
          <button
            class="rocketreserved-btn"
            type="button"
          >
            Reserve rocket
          </button>
        </div>
      </div>
      <div
        class="rocket-list-item"
      >
        <img
          class="rocket-image"
        />
        <div
          class="rocket-info"
        >
          <h2>
            Atlas V
          </h2>
          <p>
             
            A rocket developed by United Launch Alliance
          </p>
          <button
            class="rocketreserved-btn"
            type="button"
          >
            Reserve rocket
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
